runtime/protocol/http-client-engines/http-client-engine-crt/common/src/aws/smithy/kotlin/runtime/http/engine/crt/CrtHttpEngine.kt

```diff
@@ -100,6 +103,40 @@ public class CrtHttpEngine(public val config: CrtHttpEngineConfig) : HttpClientE
         val stream = conn.makeRequest(engineRequest, respHandler)
         stream.activate()
 
+        if (request.isChunked) {
+            val bytesToRead = 65536L
+
+            when (request.body) {
+                is HttpBody.SourceContent -> {
+                    val source = (request.body as HttpBody.SourceContent).readFrom()
+                    val nextBuffer = SdkBuffer()
+                    var buffer = SdkBuffer()
+
+                    while (true) {
+                        source.read(buffer, bytesToRead)
+
+                        // read another set of bytes into a second buffer -- this is how we can "peek" to decide if this is the final chunk or not
```

Comment:
**fix**: This double buffering logic is unnecessary. 

Instead you can structure this like:

```kotlin
val source = request.body.readFrom()
val bufferedSource = source.buffer()

while(true) {
    bufferedSource.request(bytesToRead)
    
    bufferedSource.buffer.readByteArray()
     ...
    val isFinalChunk = bufferedSource.exhausted()
    ...
}

```